app/scripts/controllers/transactions/index.js

```diff
@@ -798,99 +728,29 @@ export default class TransactionController extends EventEmitter {
     transactionType,
     sendFlowHistory = [],
     actionId,
+    options,
   ) {
-    if (
-      transactionType !== undefined &&
-      !VALID_UNAPPROVED_TRANSACTION_TYPES.includes(transactionType)
-    ) {
-      throw new Error(
-        `TransactionController - invalid transactionType value: ${transactionType}`,
-      );
-    }
-
-    // If a transaction is found with the same actionId, do not create a new speed-up transaction.
-    if (actionId) {
-      let existingTxMeta =
-        this.txStateManager.getTransactionWithActionId(actionId);
-      if (existingTxMeta) {
-        this.emit('newUnapprovedTx', existingTxMeta);
-        existingTxMeta = await this.addTransactionGasDefaults(existingTxMeta);
-        this._requestApproval(existingTxMeta);
-        return existingTxMeta;
-      }
-    }
-
-    // validate
-    const normalizedTxParams = txUtils.normalizeTxParams(txParams);
-    const eip1559Compatibility = await this.getEIP1559Compatibility();
-
-    txUtils.validateTxParams(normalizedTxParams, eip1559Compatibility);
-
-    /**
-     * `generateTxMeta` adds the default txMeta properties to the passed object.
-     * These include the tx's `id`. As we use the id for determining order of
-     * txes in the tx-state-manager, it is necessary to call the asynchronous
-     * method `determineTransactionType` after `generateTxMeta`.
-     */
-    let txMeta = this.txStateManager.generateTxMeta({
-      txParams: normalizedTxParams,
+    const { txMeta, isExisting } = await this._createTransaction(
+      txMethodType,
+      txParams,
       origin,
+      transactionType,
       sendFlowHistory,
-    });
-
-    // Add actionId to txMeta to check if same actionId is seen again
-    // IF request to create transaction with same actionId is submitted again, new transaction will not be added for it.
-    if (actionId) {
-      txMeta.actionId = actionId;
+      actionId,
+      options,
+    );
+    if (isExisting) {
+      return txMeta.status === TransactionStatus.unapproved
```

Comment:
From what I can see, `_waitForTransactionFinished` resolves when a `${id}:finished` event is fired.

However, that event is only fired when the status is changed to `submitted`, `rejected` or `failed`.

Since we are just awaiting if the initial status is `unapproved`, is there any chance that we could be missing out on awaiting for the tx from any of the other states (`approved`, `signed`, `dropped`, `confirmed`, `pending`). It might be the case that this is impossible, in which case we are good, but just mentioning it so that it's not overlooked.